### PR TITLE
added stokes example

### PR DESCRIPTION
**Pull request by asartori86**
this is (for now) a lid driven cavity. After refinement, it does not converge... any helps? @ESeNonFossiIo  @luca-heltai

### CODE REVIEW

**Comment by asartori86** [comment]
```
Number of active cells: 768 (on 4 levels)
Number of degrees of freedom: 7,312 (6464+848)

 iterations:                           0

 iterations:                           0
     0 ---->  0.01
     0 ---->  0.01

 iterations:                           121

 iterations:                           4

 iterations:                           130

 iterations:                           157

 iterations:                           25

 iterations:                           149

 iterations:                           6

 iterations:                           60

 iterations:                           6

 iterations:                           40

 iterations:                           7

 iterations:                           33

 iterations:                           7

 iterations:                           27

 iterations:                           8
  0.01 ---->  0.02
  0.01 ---->  0.02

 iterations:                           25

 iterations:                           24
  0.02 ---->  0.03
  0.02 ---->  0.03

 iterations:                           21

 iterations:                           22
  0.03 ---->  0.04
  0.03 ---->  0.04
  0.04 ---->  0.05
  0.04 ---->  0.05
  0.05 ---->  0.06
  0.05 ---->  0.06

 iterations:                           22

 iterations:                           22
  0.06 ---->  0.07
  0.06 ---->  0.07
  0.07 ---->  0.08
  0.07 ---->  0.08
  0.08 ---->  0.09
  0.08 ---->  0.09
  0.09 ---->   0.1
  0.09 ---->   0.1
   0.1 ---->  0.11
   0.1 ---->  0.11

 iterations:                           23
  0.11 ---->  0.12
  0.11 ---->  0.12
  0.12 ---->  0.13
  0.12 ---->  0.13
  0.13 ---->  0.14
  0.13 ---->  0.14
  0.14 ---->  0.15
  0.14 ---->  0.15
  0.15 ---->  0.16
  0.15 ---->  0.16

 iterations:                           23

 iterations:                           23
  0.16 ---->  0.17
  0.16 ---->  0.17
  0.17 ---->  0.18
  0.17 ---->  0.18
  0.18 ---->  0.19
  0.18 ---->  0.19
  0.19 ---->   0.2
  0.19 ---->   0.2
   0.2 ---->  0.21
   0.2 ---->  0.21
  0.21 ---->  0.22
  0.21 ---->  0.22
  0.22 ---->  0.23
  0.22 ---->  0.23
  0.23 ---->  0.24
  0.23 ---->  0.24
  0.24 ---->  0.25
  0.24 ---->  0.25
  0.25 ---->  0.26
  0.25 ---->  0.26

 iterations:                           23
  0.26 ---->  0.27
  0.26 ---->  0.27
  0.27 ---->  0.28
  0.27 ---->  0.28
  0.28 ---->  0.29
  0.28 ---->  0.29
  0.29 ---->   0.3
  0.29 ---->   0.3
   0.3 ---->  0.31
   0.3 ---->  0.31
  0.31 ---->  0.32
  0.31 ---->  0.32
  0.32 ---->  0.33
  0.32 ---->  0.33
  0.33 ---->  0.34
  0.33 ---->  0.34
  0.34 ---->  0.35
  0.34 ---->  0.35
  0.35 ---->  0.36
  0.35 ---->  0.36

 iterations:                           23
  0.36 ---->  0.37
  0.36 ---->  0.37
  0.37 ---->  0.38
  0.37 ---->  0.38
  0.38 ---->  0.39
  0.38 ---->  0.39
  0.39 ---->   0.4
  0.39 ---->   0.4
   0.4 ---->  0.41
   0.4 ---->  0.41
  0.41 ---->  0.42
  0.41 ---->  0.42
  0.42 ---->  0.43
  0.42 ---->  0.43
  0.43 ---->  0.44
  0.43 ---->  0.44
  0.44 ---->  0.45
  0.44 ---->  0.45
  0.45 ---->  0.46
  0.45 ---->  0.46
  0.46 ---->  0.47
  0.46 ---->  0.47

 iterations:                           24

 iterations:                           24
  0.47 ---->  0.48
  0.47 ---->  0.48
  0.48 ---->  0.49
  0.48 ---->  0.49
  0.49 ---->   0.5
  0.49 ---->   0.5
   0.5 ---->  0.51
   0.5 ---->  0.51
  0.51 ---->  0.52
  0.51 ---->  0.52
  0.52 ---->  0.53
  0.52 ---->  0.53
  0.53 ---->  0.54
  0.53 ---->  0.54
  0.54 ---->  0.55
  0.54 ---->  0.55
  0.55 ---->  0.56
  0.55 ---->  0.56
  0.56 ---->  0.57
  0.56 ---->  0.57
  0.57 ---->  0.58
  0.57 ---->  0.58
  0.58 ---->  0.59
  0.58 ---->  0.59
  0.59 ---->   0.6
  0.59 ---->   0.6
   0.6 ---->  0.61
   0.6 ---->  0.61
  0.61 ---->  0.62
  0.61 ---->  0.62
  0.62 ---->  0.63
  0.62 ---->  0.63
  0.63 ---->  0.64
  0.63 ---->  0.64
  0.64 ---->  0.65
  0.64 ---->  0.65
  0.65 ---->  0.66
  0.65 ---->  0.66
  0.66 ---->  0.67
  0.66 ---->  0.67

 iterations:                           25
  ################ restart ######### 
max_kelly > threshold
0.0101978 >  0.01
######################################
Number of active cells: 258 (on 5 levels)
Number of degrees of freedom: 2,715 (2392+323)

 iterations:                           187
par
 iterations:                           517

 iterations:                           684

 iterations:                           875

 iterations:                           986
```

**Comment by asartori86** [comment]
video

**Comment by nicola-giuliani** [comment]
It should not help the convergence but it seems to me that there is not any manifold_id for the hole

File: examples/stokes/source/stokes_ida.cc [hunk 1]
```diff
@@ -0,0 +1,858 @@
...
+                  "1.",
+                  Patterns::Double(0.0));
+}
+
+template <int dim>
+void Stokes<dim>::make_grid_fe()
+{
+  triangulation = SP(pgg.distributed(comm));
+  dof_handler = SP(new DoFHandler<dim>(*triangulation));
+  fe=SP(fe_builder());
```
Comment (nicola-giuliani) [inline examples/stokes/source/stokes_ida.cc]: If we want the cylindrical hole we need to set up the manifold here. The only way I know is with something like this

const Point<2> center (0,0);
const double inner_radius = 0.2;

  const SphericalManifold<2> manifold_description(center);
  Triangulation<2>::active_cell_iterator
  cell = triangulation.begin_active(),
  endc = triangulation.end();
  for (; cell!=endc; ++cell)
    {
      for (unsigned int v=0;
           v < GeometryInfo<2>::faces_per_cell;
           ++v)
        {
               const double distance_from_center
        = center.distance (cell->face(v)->center());
        if (std::fabs(distance_from_center) < inner_radius + 1e-6)
          {
            cell->face(v)->set_manifold_id (99);
            break;
}}
triangulation.set_manifold (99, manifold_description);

**Comment by asartori86** [comment]
```
alberto@lenovo build (stokes_example *)$ mpirun -np 2 ./stokes.g 
Number of active cells: 12 (on 1 levels)
Number of degrees of freedom: 158 (136+22)

 iterations:                           0

 iterations:                           0
     0 ---->  0.01
     0 ---->  0.01

 iterations:                           46

 iterations:                           2

 iterations:                           48

 iterations:                           53

 iterations:                           12

 iterations:                           60

 iterations:                           3

 iterations:                           20

 iterations:                           3

 iterations:                           29

 iterations:                           3

 iterations:                           27

 iterations:                           3

 iterations:                           22

 iterations:                           3
  0.01 ---->  0.02
  0.01 ---->  0.02

 iterations:                           26

 iterations:                           27
  0.02 ---->  0.03
  0.02 ---->  0.03

 iterations:                           26

 iterations:                           25
  0.03 ---->  0.04
  0.03 ---->  0.04
  0.04 ---->  0.05
  0.04 ---->  0.05
  0.05 ---->  0.06
  0.05 ---->  0.06

 iterations:                           22

 iterations:                           22
  0.06 ---->  0.07
  0.06 ---->  0.07
  0.07 ---->  0.08
  0.07 ---->  0.08
  0.08 ---->  0.09
  0.08 ---->  0.09
  0.09 ---->   0.1
  0.09 ---->   0.1
   0.1 ---->  0.11
   0.1 ---->  0.11

 iterations:                           23
  0.11 ---->  0.12
  0.11 ---->  0.12
  0.12 ---->  0.13
  0.12 ---->  0.13
  0.13 ---->  0.14
  0.13 ---->  0.14
  0.14 ---->  0.15
  0.14 ---->  0.15
  0.15 ---->  0.16
  0.15 ---->  0.16

 iterations:                           22

 iterations:                           22
  0.16 ---->  0.17
  0.16 ---->  0.17
  0.17 ---->  0.18
  0.17 ---->  0.18
  0.18 ---->  0.19
  0.18 ---->  0.19
  0.19 ---->   0.2
  0.19 ---->   0.2
   0.2 ---->  0.21
   0.2 ---->  0.21
  0.21 ---->  0.22
  0.21 ---->  0.22
  0.22 ---->  0.23
  0.22 ---->  0.23
  0.23 ---->  0.24
  0.23 ---->  0.24
  0.24 ---->  0.25
  0.24 ---->  0.25
  0.25 ---->  0.26
  0.25 ---->  0.26

 iterations:                           23

 iterations:                           23
  ################ restart ######### 
max_kelly > threshold
0.0110894 >  0.01
######################################
Number of active cells: 18 (on 2 levels)
Number of degrees of freedom: 240 (208+32)

 iterations:                           75

 iterations:                           94

 iterations:                           139

 iterations:                           160

 iterations:                           214

[IDA ERROR]  IDACalcIC

[IDA ERROR]  IDACalcIC
  The residual routine or the linear setup or solve routine had a recoverable error, but IDACalcIC was unable to recover.

  The residual routine or the linear setup or solve routine had a recoverable error, but IDACalcIC was unable to recover.

  ################ restart ######### 
max_kelly > threshold
0.0182418 >  0.01
######################################
Number of active cells: 24 (on 2 levels)
Number of degrees of freedom: 295 (256+39)

 iterations:                           26

 iterations:                           70

 iterations:                           119

 iterations:                           155

 iterations:                           209

[IDA ERROR]  IDACalcIC
  The residual routine or the linear setup or solve routine had a recoverable error, but IDACalcIC was unable to recover.


[IDA ERROR]  IDACalcIC
  The residual routine or the linear setup or solve routine had a recoverable error, but IDACalcIC was unable to recover.

  ################ restart ######### 
max_kelly > threshold
0.0185276 >  0.01
######################################
Number of active cells: 33 (on 3 levels)
Number of degrees of freedom: 415 (362+53)

 iterations:                           33

 iterations:                           89

 iterations:                           155
```

**Comment by asartori86** [comment]
video_no_bc

**Comment by asartori86** [comment]
se non raffino va via liscio

**Comment by asartori86** [comment]
rotta

**Comment by asartori86** [comment]
se raffino globalmente ora ce la fa. ma secondo me non funziona il parallel::distributed::adaptive

**Comment by luca-heltai** [comment]
Il problema quindi e' in parallel distributed adaptive, oppure no?

**Comment by asartori86** [comment]
parallell::distributed::adaptive sembra non vedere la manifold. però anche se tolgo la manifold, una volta raffinato localmente non riesco più a ripartire

**Comment by luca-heltai** [comment]
quindi questo

   parallell::distributed::adaptive sembra non vedere la manifold

e' un baco di deal.II?

**Comment by asartori86** [comment]
direi di si

**Comment by luca-heltai** [comment]
Anche su un solo processore?

**Comment by asartori86** [comment]
con un solo processore la mesh sta benone

**Comment by luca-heltai** [comment]
Mini test da postare su dealii bugs?

**Comment by luca-heltai** [comment]
fai la mesh, assegni il manifold, marchi un po' di celle a caso per il raffinamento,  e vedi che succede con uno e due proccesori

**Comment by asartori86** [comment]
works

**Comment by luca-heltai** [comment]
ping @asartori86 

What's the status of the Stokes/Navier-Stokes example? Did we understand what happened with the linear operator ?

File: examples/stokes/source/stokes_ida.cc [hunk 2]
```diff
@@ -0,0 +1,976 @@
...
+        cell->get_dof_indices (local_dof_indices);
+        constraints.distribute_local_to_global (cell_matrix,
+                                                local_dof_indices,
+                                                jacobian_matrix);
+
+        constraints.distribute_local_to_global (cell_prec,
+                                                local_dof_indices,
+                                                jacobian_preconditioner_matrix);
+
+     //   constraints_dot.distribute_local_to_global (cell_matrix,
```
Comment (luca-heltai) [inline examples/stokes/source/stokes_ida.cc]: remove these comments

File: examples/stokes/source/stokes_ida.cc [hunk 3]
```diff
@@ -0,0 +1,976 @@
...
+
+     //   constraints_dot.distribute_local_to_global (cell_prec,
+     //                                           local_dof_indices,
+     //                                           jacobian_preconditioner_matrix);
+      }
+
+  jacobian_matrix.compress (VectorOperation::add);
+  jacobian_preconditioner_matrix.compress (VectorOperation::add);
+  //compress(jacobian_matrix, VectorOperation::add);
+/* mbarde
```
Comment (luca-heltai) [inline examples/stokes/source/stokes_ida.cc]: Also these comments

File: examples/stokes/source/stokes_ida.cc [hunk 4]
```diff
@@ -0,0 +1,976 @@
...
+      double mpi_max_kelly=0;
+
+      computing_timer.enter_section ("   Compute error estimator");
+
+
+  update_constraints(t);
+
+     VEC tmp_c(solution);
+     constraints.distribute(tmp_c);
+     distributed_solution = tmp_c;
```
Comment (luca-heltai) [inline examples/stokes/source/stokes_ida.cc]: indentation check should fail here...

File: examples/stokes/source/stokes_ida.cc [hunk 5]
```diff
@@ -0,0 +1,976 @@
...
+          const LinearSolvers::BlockSchurPreconditioner<TrilinosWrappers::PreconditionAMG,
+                TrilinosWrappers::PreconditionJacobi>
+                preconditioner (jacobian_matrix, jacobian_preconditioner_matrix,
+                                *Mp_preconditioner, *Amg_preconditioner,
+                                false);
+
+          SolverFGMRES<TrilinosWrappers::MPI::BlockVector>
+          solver(solver_control, mem,
+                 SolverFGMRES<TrilinosWrappers::MPI::BlockVector>::
+                 AdditionalData(30, true));
```
Comment (luca-heltai) [inline examples/stokes/source/stokes_ida.cc]: Shouldn't this be a false?

File: examples/stokes/source/stokes_ida.cc [hunk 6]
```diff
@@ -0,0 +1,976 @@
...
+
+template <int dim>
+VEC &Stokes<dim>::differential_components() const
+{
+  static VEC diff_comps;
+  diff_comps.reinit(solution);
+  diff_comps.block(0) = 1; // velocity is differential
+  diff_comps.block(1) = 0; // pressure is only algebraic
+
+ // set_constrained_dofs_to_zero(diff_comps);
```
Comment (luca-heltai) [inline examples/stokes/source/stokes_ida.cc]: Why is this commented? Constrained degrees of freedom should be algebraic...

File: examples/stokes/source/stokes_ida.cc.bak [hunk 7]
```diff
@@ -0,0 +1,951 @@
+//-----------------------------------------------------------
```
Comment (luca-heltai) [inline examples/stokes/source/stokes_ida.cc.bak]: What is this file doing here?

File: examples/stokes/step-1.cc [hunk 8]
```diff
@@ -0,0 +1,221 @@
+/* ---------------------------------------------------------------------
```
Comment (luca-heltai) [inline examples/stokes/step-1.cc]: And what is this one doing here?

**Comment by asartori86** [comment]
@luca-heltai @ESeNonFossiIo everything works, except for the solver using the linear operator..

**Pull request merged**